pkg/storage/chunk/prometheus_chunk.go

```diff
@@ -106,16 +162,22 @@ func (p *prometheusChunkIterator) Value() model.SamplePair {
 	}
 }
 
-func (p *prometheusChunkIterator) Batch(size int) Batch {
+// TODO native histograms support, assumes it's used on float chunk and only keeps floats
+func (p *prometheusChunkIterator) Batch(size int, valueType chunkenc.ValueType) Batch {
 	var batch Batch
 	j := 0
 	for j < size {
-		t, v := p.it.At()
-		batch.Timestamps[j] = t
-		batch.Values[j] = v
-		j++
-		if j < size && !p.it.Next() {
-			break
+		if valueType == chunkenc.ValFloat {
+			t, v := p.it.At()
+			batch.Timestamps[j] = t
+			batch.Values[j] = v
+			j++
+		}
+		if j < size {
+			valueType = p.it.Next()
```

Comment:
You're overwriting the input `valueType` which looks wrong to me. To my understanding we're not using the input `valueType` yet, but we should rename it to avoid future bugs.

Reply:
Actually I'm not sure my comment is correct. `valueType` is the type of the current value (like the one returned by the previous call to `Next()`). If this comment is correct, then I would add a nice comment to `Batch()` to explain it.

Reply:
It turns out that we process the first sample _before_ we call `Next` , so we need to know it's type already, hence the type (for first sample) is given in the signature. Adding comment.